### PR TITLE
Update cohort data generation to  read concept name from concept table

### PR DESCRIPTION
**Pull request by tianj7**
Ticket: https://ctds-planx.atlassian.net/browse/VADC-847

Grab hare columns from file instead of hardcoded value

### Improvements
Read concept name from concept table when creating cohort data file

### CODE REVIEW

File: models/cohortdata.go [hunk 1]
```diff
@@ -83,9 +83,10 @@ func (h CohortData) RetrieveDataBySourceIdAndCohortIdAndConceptIdsOrderedByPerso
...
 	var cohortData []*PersonConceptAndValue
 	query := omopDataSource.Db.Table(omopDataSource.Schema+".observation_continuous as observation"+omopDataSource.GetViewDirective()).
-		Select("observation.person_id, observation.observation_concept_id as concept_id, concept.concept_class_id, observation.value_as_string as concept_value_as_string, observation.value_as_number as concept_value_as_number, observation.value_as_concept_id as concept_value_as_concept_id").
+		Select("observation.person_id, observation.observation_concept_id as concept_id, concept.concept_class_id, hare_concept.concept_name as concept_value_name, observation.value_as_number as concept_value_as_number, observation.value_as_concept_id as concept_value_as_concept_id").
 		Joins("INNER JOIN "+resultsDataSource.Schema+".cohort as cohort ON cohort.subject_id = observation.person_id").
 		Joins("INNER JOIN "+omopDataSource.Schema+".concept as concept ON concept.concept_id = observation.observation_concept_id").
+		Joins("LEFT JOIN "+omopDataSource.Schema+".concept as hare_concept ON hare_concept.concept_id = observation.value_as_concept_id").
```
Comment (pieterlukasse) [inline models/cohortdata.go:89]:
```suggestion
		Joins("LEFT JOIN "+omopDataSource.Schema+".concept as value_as_concept ON value_as_concept.concept_id = observation.value_as_concept_id").
```
I think `value_as_concept` is a better, more generic name.

File: models/cohortdata.go [hunk 2]
```diff
@@ -83,9 +83,10 @@ func (h CohortData) RetrieveDataBySourceIdAndCohortIdAndConceptIdsOrderedByPerso
 	// get the observations for the subjects and the concepts, to build up the data rows to return:
 	var cohortData []*PersonConceptAndValue
 	query := omopDataSource.Db.Table(omopDataSource.Schema+".observation_continuous as observation"+omopDataSource.GetViewDirective()).
-		Select("observation.person_id, observation.observation_concept_id as concept_id, concept.concept_class_id, observation.value_as_string as concept_value_as_string, observation.value_as_number as concept_value_as_number, observation.value_as_concept_id as concept_value_as_concept_id").
+		Select("observation.person_id, observation.observation_concept_id as concept_id, concept.concept_class_id, hare_concept.concept_name as concept_value_name, observation.value_as_number as concept_value_as_number, observation.value_as_concept_id as concept_value_as_concept_id").
```
Comment (pieterlukasse) [inline models/cohortdata.go:86]:
```suggestion
		Select("observation.person_id, observation.observation_concept_id as concept_id, concept.concept_class_id, value_as_concept.concept_name as concept_value_name, observation.value_as_number as concept_value_as_number, observation.value_as_concept_id as concept_value_as_concept_id").
```

File: tests/models_tests/models_test.go [hunk 3]
```diff
@@ -814,18 +814,27 @@ func TestRetrieveDataBySourceIdAndCohortIdAndConceptIdsOrderedByPersonId(t *test
 			t.Errorf("Expected some cohort data")
 		}
 		var previousPersonId int64 = -1
+		defaultHareValue := map[string]bool{"": true, "non-Hispanic Asian": true, "non-Hispanic Black": true, "non-Hispanic White": true, "Hispanic": true}
```
Comment (pieterlukasse) [inline tests/models_tests/models_test.go:817]: why is `"": true` ?

File: models/cohortdata.go [hunk 4]
```diff
@@ -21,7 +21,7 @@ type PersonConceptAndValue struct {
 	PersonId                int64
 	ConceptId               int64
 	ConceptClassId          string
-	ConceptValueAsString    string
+	ConceptValueName        string
```
Comment (pieterlukasse) [inline models/cohortdata.go:24]:
```suggestion
	ConceptValueAsConceptName        string
```
although longer, I think this better aligns with the naming for the other variables (see the ones just below).

File: tests/models_tests/models_test.go (hunk 3 above)
Comment (tianj7) [reply tests/models_tests/models_test.go:817]: There is a hare concept observation row with empty value_as_concept_id in the test data set.

Comment (tianj7) [reply tests/models_tests/models_test.go:817]: https://github.com/uc-cdis/cohort-middleware/blob/944e952a97fd3085801a32c9781da0fab81b59a9/tests/setup_local_db/test_data_results_and_cdm.sql#L119

Comment (tianj7) [reply tests/models_tests/models_test.go:817]: I modified the test to only allow empty value once.

File: models/cohortdata.go (hunk 4 above)
Comment (tianj7) [reply models/cohortdata.go:24]: I think this name is more confusing actually to me than ConceptValueName. The Observation already have a concept, but this name is the observation value, which itself is another concept. Therefore concept value('s) name

If we really need to make it more specific, maybe ObservationValueConceptName?

Comment (pieterlukasse) [reply models/cohortdata.go:24]: Ok, I agree. I like the one you proposed, with a little extra: ObservationValue**As**ConceptName

File: tests/models_tests/models_test.go [hunk 5]
```diff
@@ -814,17 +814,22 @@ func TestRetrieveDataBySourceIdAndCohortIdAndConceptIdsOrderedByPersonId(t *test
 			t.Errorf("Expected some cohort data")
 		}
 		var previousPersonId int64 = -1
-		defaultHareValue := map[string]bool{"": true, "non-Hispanic Asian": true, "non-Hispanic Black": true, "non-Hispanic White": true, "Hispanic": true}
-
+		defaultHareValue := map[string]bool{"non-Hispanic Asian": true, "non-Hispanic Black": true, "non-Hispanic White": true, "Hispanic": true}
+		emptyDataCounter := 0
```
Comment (pieterlukasse) [inline tests/models_tests/models_test.go:818]: While testing locally I noticed cohortData exploded. I think the query might be wrong. Please check.
Please also add this check to complement the `// 1- ` check above and guard against regression:
```go
		// 2- check data size. Max size, if all persons have data for each concept, is cohort size x len(allConceptIds):
		if len(cohortData) > cohortDefinition.CohortSize*len(allConceptIds) {
			t.Errorf("Cohort data size does not match expectation. Expected < %d, got %d",
				cohortDefinition.CohortSize*len(allConceptIds), len(cohortData))
		}
```

Comment (pieterlukasse) [reply tests/models_tests/models_test.go:824]: issue could not be reproduced. But test/check added to make sure.

**Pull request merged**